ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,559 @@
+	Roots     map[*node.Node]uint64 // node index for root nodes
+	NodeCount uint64                // total number of children nodes for the subtrie
+	Checksum  uint32                // checksum of written file
+	Err       error                 // indicate whether storing subtrie nodes was running into any error
+}
+
+// StoreCheckpointV6 stores checkpoint file into a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+// any error returned are exception
+func StoreCheckpointV6(
+	tries []*trie.MTrie, outputDir string, outputFile string, logger *zerolog.Logger) error {
+	if len(tries) == 0 {
+		logger.Info().Msg("no tries to be checkpointed")
+		return nil
+	}
+
+	first, last := tries[0], tries[len(tries)-1]
+	logger.Info().
+		Str("first_hash", first.RootHash().String()).
+		Uint64("first_reg_count", first.AllocatedRegCount()).
+		Str("last", last.RootHash().String()).
+		Uint64("last_reg_count", last.AllocatedRegCount()).
+		Msgf("storing checkpoint for %v tries to %v", len(tries), outputDir)
+
+	subtrieRoots := createSubTrieRoots(tries)
+
+	estimatedSubtrieNodeCount := estimateSubtrieNodeCount(tries)
+
+	subTrieRootIndices, subTriesNodeCount, subTrieChecksums, err := storeSubTrieConcurrently(
+		subtrieRoots,
+		estimatedSubtrieNodeCount,
+		outputDir,
+		outputFile,
+		logger,
+	)
+	if err != nil {
+		return fmt.Errorf("could not store sub trie: %w", err)
+	}
+
+	logger.Info().Msgf("subtrie have been stored. sub trie node count: %v", subTriesNodeCount)
+
+	topTrieChecksum, err := storeTopLevelNodesAndTrieRoots(
+		tries, subTrieRootIndices, subTriesNodeCount, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store top level tries: %w", err)
+	}
+
+	err = storeCheckpointSummary(subTrieChecksums, topTrieChecksum, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+
+	logger.Info().Msgf("checkpoint file has been successfully stored at %v/%v", outputDir, outputFile)
+
+	return nil
+}
+
+// storeCheckpointSummary stores the checkpoint file header, which contains the following information
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// any error returned are exception
+func storeCheckpointSummary(
+	subTrieChecksums []uint32,
+	topTrieChecksum uint32,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) error {
+	closable, err := createWriterForCheckpointSummary(outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
```

Comment:
instead of closing closable here, can you close the crc writer instead?

```
        checksum := writer.Crc32()
	_, err = writer.Write(encodeCRC32Sum(checksum))
	if err != nil {
		return fmt.Errorf("cannot write CRC32 checksum to checkpoint summary: %w", err)
	}
```
Should be part of crc writer's Close method

Reply:
`Crc32Writer` wraps a io.Writer interface, which doesn't have `Close` method, so itself also don't have `Close` method. In other words, `Crc32Writer` doesn't care whether the writer is closable or not.

Reply:
we own Crc32Writer, so why can't we change it to take a io.WriteCloser instead of just a io.Writer?

Reply:
Thanks for the idea, [tracked here for now](https://github.com/onflow/flow-go/issues/3323)